File: JumpGameDFS.java
import java.util.Scanner;
import java.util.Arrays;

public class JumpGameDFS {

        // DFS - Time Exponential - Space O(n)

        int n;
        int[] visited;
        int[] nums;

        public boolean canJump(int[] nums) {

            this.nums = nums;
            n = nums.length ;

            // single element in nums
            if(n < 2) {
                return true;
            }

            visited = new int[n];

            // -1 -> not visited, 1 -> visited, 2 -> conclusively does not lead to the target
            // all elements are unvisited by default
            Arrays.fill(visited, -1);

            // run dfs from zeroth index
            return dfs(0);
        }


        public boolean dfs(int idx) {

            // if index conclusively leads to the target by surpassing the last index of nums, success
            if(idx >= n-1) {
                return true;
            }

            // if conclusively does not lead to the target, failure
            if(visited[idx] == 2) {
                return false;
            }

            // make it visited
            visited[idx] = 1;

            // allowed jumps from index depth wise
            for(int j = nums[idx]; j >= 1; j--) {

                // new index after jump
                int newIdx = j + idx;

                // if new index surpasses the last index of nums, success
                if(newIdx >= n-1) {
                    return true;
                }

                // if unvisited index and makes dfs true, success
                if(visited[newIdx] == -1 && dfs(newIdx)) {
                    return true;
                }

            }
            // after allowed jumps with no success, target is unreachable
            visited[idx] = 2;
            return false;
        }

        public static void main(String[] args) {

            JumpGameDFS obj = new JumpGameDFS();
            Scanner scanner = new Scanner(System.in);

            System.out.println("array length: ");
            int n = scanner.nextInt();

            int[] nums = new int[n];
            System.out.println("array: ");
            for(int i = 0; i < n; i++) {
                nums[i] = scanner.nextInt();
            }

            boolean answer = obj.canJump(nums);
            System.out.println("can the last index be reached? " + answer);
        }
}

/*
TIME COMPLEXITY = Exponential

SPACE COMPLEXITY = O(n)

Recursive stack space = O(n)
visited array = O(n)
*/

File: Jump_game_DFS.py
# DFS
# Time complexity : O(n^k)
# Space complexity : O(n^k)
# Leetcode : Time limit exceeded

from collections import deque
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # if the list has length less than 2, return True
        if len(nums) < 2:
            return True
        
        # find the length of the list and call the dfs function with the starting index as 0
        n = len(nums)
        return self.dfs(nums, 0)
    
    def dfs(self, nums, i):
        # if we have reached or go over the lenghth of list or the last index, then return True
        if i >= len(nums) - 1:
            return True
        
        # start the traversal from the last element till 0 and recursively call the dfs function on the nums array and the new index
        for j in range(nums[i],0,-1):
            idx =  i + j
            if self.dfs(nums, idx):
                return True
        
        return False
